### PR TITLE
tests for methods that hit the templates endpoint

### PR DESCRIPTION
**Pull request by yargevad**
Here are some helpers for visualizing code coverage locally:

```
function gocover () {
  dir=${PWD##*/}
  day=$(date +"%Y%m%d")
  t="/tmp/${dir}.${day}"
  go test -race -coverprofile="$t" $@
  go tool cover -html="$t" -o="$t.html"
}

function gocovero () {
  dir=${PWD##*/}
  day=$(date +"%Y%m%d")
  t="/tmp/${dir}.${day}"
  go test -race -coverprofile="$t" $@
  go tool cover -html="$t" -o="$t.html"
  open "$t.html"
}
```

### CODE REVIEW

File: common.go [hunk 1]
```diff
@@ -99,43 +96,11 @@ func (api *Client) Init(cfg *Config) error {
 		cfg.ApiVersion = 1
 	}
 	api.Config = cfg
-	api.headers = make(map[string]string)
-
-	if api.Client == nil {
```
Comment (yargevad) [inline common.go]: Removing this block - callers can supply their own `http.Client` with the same sort of workaround if necessary.

File: common.go [hunk 2]
```diff
@@ -31,10 +28,12 @@ type Config struct {
 // Client contains connection, configuration, and authentication information.
 // Specifying your own http.Client gives you lots of control over how connections are made.
 // Clients are safe for concurrent (read-only) reuse by multiple goroutines.
+// Headers is useful to set subaccount (X-MSYS-SUBACCOUNT header) and any other custom headers.
+// All calls to SetHeader must happen before Client is exposed to possible concurrent use.
 type Client struct {
 	Config  *Config
 	Client  *http.Client
-	headers map[string]string
+	Headers *http.Header
```
Comment (yargevad) [inline common.go]: Export this and take advantage of the `http.Header` methods

File: common.go [hunk 3]
```diff
@@ -99,43 +96,11 @@ func (api *Client) Init(cfg *Config) error {
...
-func (c *Client) SetHeader(header string, value string) {
-	c.headers[header] = value
-}
-
-// RemoveHeader removes a header set in SetHeader function
-// All calls to RemoveHeader must happen before Client is exposed to possible concurrent use.
-func (c *Client) RemoveHeader(header string) {
-	delete(c.headers, header)
-}
-
```
Comment (yargevad) [inline common.go]: `http.Header` methods are better, remove these functions in favor of those

File: common.go [hunk 4]
```diff
@@ -206,15 +173,9 @@ func (c *Client) DoRequest(ctx context.Context, method, urlStr string, data []by
 	// set any headers provided in context
 	if header, ok := ctx.Value("http.Header").(http.Header); ok {
 		for key, vals := range map[string][]string(header) {
-			if len(vals) >= 1 {
-				// replace existing headers, default, or from Client.headers
-				req.Header.Set(key, vals[0])
-			}
-			if len(vals) > 2 {
```
Comment (yargevad) [inline common.go]: Simplifying fixes the off-by-one error that was here, where the second header wouldn't be applied to the request, but the first and third etc would.

File: common.go [hunk 5]
```diff
@@ -234,7 +195,7 @@ func (c *Client) DoRequest(ctx context.Context, method, urlStr string, data []by
 	if c.Config.Verbose {
 		ares.Verbose["http_status"] = ares.HTTP.Status
 		bodyBytes, dumpErr := httputil.DumpResponse(res, true)
-		if err != nil {
+		if dumpErr != nil {
```
Comment (yargevad) [inline common.go]: Still need to read up on when this error condition can be triggered.

File: common_test.go [hunk 6]
```diff
@@ -53,3 +54,68 @@ func testFailVerbose(t *testing.T, res *sp.Response, fmt string, args ...interfa
...
+	if str != exp {
+		t.Errorf("*Error.Json => %q, want %q", str, exp)
+	}
+}
+
+func TestDoRequest_BadMethod(t *testing.T) {
+	testSetup(t)
+	defer testTeardown()
+
+	_, err := testClient.DoRequest(nil, "💩", "", nil)
```
Comment (yargevad) [inline common_test.go]: This makes me so happy.

File: subaccounts.go [hunk 7]
```diff
@@ -153,7 +153,7 @@ func (c *Client) SubaccountUpdateContext(ctx context.Context, s *Subaccount) (re
 		return
 	}
 
-	path := fmt.Sprintf(templatesPathFormat, c.Config.ApiVersion)
```
Comment (yargevad) [inline subaccounts.go]: This method was busted, hitting the wrong endpoint

File: common_test.go (hunk 6 above)
Comment (yepher) [inline common_test.go]: cool, nice use of emoji.. I did not know you could do that.

**Pull request merged**